pkg/daemon/daemon.go

```diff
@@ -1093,14 +1167,12 @@ func (dn *Daemon) checkStateOnFirstRun() error {
 		expectedConfig = state.currentConfig
 	}
 
-	if _, err := os.Stat(constants.MachineConfigDaemonForceFile); err != nil {
-		if err := dn.validateOnDiskState(expectedConfig); err != nil {
-			return fmt.Errorf("unexpected on-disk state validating against %s: %v", expectedConfig.GetName(), err)
-		}
-		glog.Info("Validated on-disk state")
-	} else {
-		glog.Infof("Skipping on-disk validation; %s present", constants.MachineConfigDaemonForceFile)
+	if skipped, err := dn.maybeValidateOnDiskState(expectedConfig); skipped {
+		// We skipped validation, so immediately trigger the update
```

Comment:
Ah so this answers my question from the demo, I see you adapted the original behaviour of always triggering an update here. Reading the code it appears that only on checkStateOnFirstRun do we always force an upgrade if a forcefile exists, but going through the logic, we appear to "apply" a change even if current==desired, and then... write the desired configuration mostly as it stands without checking for diff (e.g. re-write all the files, comparing the OSURL of the desired NOT with the current but with the BOOTED url). Then... reboot because the postConfigChangeAction honours the forcefile still.

So then we actually CAN apply a no-diff update (I didn't think this was the case, but it appears to be) and barring some subfields (e.g. kargs since that works on a diff check system) this mostly works out (at least I think so based on my understanding of the logic).

Hmm, regardless, we seem to be somewhat inconsistent here. Happy to discuss this further, I could be completely mistaken as well on some of my assumptions.